mailbox/backup/src/main/java/org/apache/james/mailbox/backup/DefaultMailboxBackup.java

```diff
@@ -0,0 +1,167 @@
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class DefaultMailboxBackup implements MailboxBackup {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultMailboxBackup.class);
+
+    protected class MailAccountContent {
+        private final MailboxWithAnnotations mailboxWithAnnotations;
+        private final Stream<MessageResult> messages;
+
+        MailAccountContent(MailboxWithAnnotations mailboxWithAnnotations, Stream<MessageResult> messages) {
+            this.mailboxWithAnnotations = mailboxWithAnnotations;
+            this.messages = messages;
+        }
+
+        public MailboxWithAnnotations getMailboxWithAnnotations() {
+            return mailboxWithAnnotations;
+        }
+
+        public Stream<MessageResult> getMessages() {
+            return messages;
+        }
+    }
+
+    public DefaultMailboxBackup(MailboxManager mailboxManager, ArchiveService archiveService, MailArchivesLoader archiveLoader, BlobStore store) {
+        this.mailboxManager = mailboxManager;
+        this.archiveService = archiveService;
+        this.archiveLoader = archiveLoader;
+        this.store = store;
+    }
+
+    @Override
+    public Publisher<BlobId> backupAccount(User user) throws IOException, MailboxException {
+        MailboxSession session = mailboxManager.createSystemSession(user.asString());
+        List<MailAccountContent> accountContents = getAccountContentForUser(session);
+        List<MailboxWithAnnotations> mailboxes = accountContents.stream()
+            .map(MailAccountContent::getMailboxWithAnnotations)
+            .collect(Guavate.toImmutableList());
+
+        Stream<MessageResult> messages = allMessagesForUser(accountContents);
+        return saveToStore(mailboxes, messages);
+    }
+
+    @Override
+    public Publisher<Void> restore(User user, BlobId blobId) throws IOException, MailboxException {
+        MailboxSession session = mailboxManager.createSystemSession(user.asString());
+        org.apache.james.blob.api.BlobId.Factory blobIdFactory = new HashBlobId.Factory();
+        try (InputStream in = store.read(blobIdFactory.from(blobId.asString()));
+             MailArchiveIterator archiveIterator = archiveLoader.load(in)) {
+            throw new NotImplementedException("TODO");
+        }
+    }
+
+    private final MailboxManager mailboxManager;
+    private final ArchiveService archiveService;
+    private final MailArchivesLoader archiveLoader;
+    private final BlobStore store;
+
+    private Function<MailboxPath, Optional<MailAccountContent>> getMailboxWithAnnotationsFromPath(MailboxSession session) {
+        return path -> {
+            try {
+                StoreMessageManager messageManager = (StoreMessageManager) mailboxManager.getMailbox(path, session);
+                Mailbox mailbox = messageManager.getMailboxEntity();
```

Comment:
Ok, the cast is needed for this...

I have a big deal with that, the backup API is mixing several level of models: the `mailbox-api` level as well as the mapper level. Which is BAD.

Not your fault, but I do belive this is more than the time to promote `Mailbox` interface + `SimpleMailbox` class as part of `mailbox-api`. That way, getMailboxEntity could be a generic method of MessageManager & we will stop this casting/abstraction level mess.